### DISCUSSION TITLE
Wrapping questions

### DISCUSSION THREAD

**Discussion by supsm** [ptitSeb/box64#799] [Q&A]
Is it possible to wrap a function with a struct argument in box64? My first attempts with the wrapper helper didn't work since it says that everything which isn't a builtin type, void, pointer, integer, character, or floating point is unsupported (see [gen.cpp:175](https://github.com/ptitSeb/box64/blob/main/wrapperhelper/gen.cpp#L175)). I also couldn't find an existing wrapped library that has this, though frankly I just don't want to dig through hundreds of files.
Additionally, I am a little confused as to the differences of the `xxx.c` and `xxx_private.h` files. Perhaps there is a guide or documentation on wrapping somewhere?

**Comment by ksco** [comment]
If the `sizeof` the struct is small enough and the position is at the front, so it is passed through the registers, you can use two builtin types to simulate. If the struct has to be passed entirely on the stack (on both sides!), then it's just the same as Vaargs (A pointer to somewhere), and you can use `V` to simulate. Other cases are not yet supported.

`xxx_private.h` is used to put the function signatures, box64 needs this information when calling native functions. Some more complex functions need more care, such as functions with callbacks or vaargs. And `xxx.c` is used to put these complex function wrappers (starting with `my_`).

**Reply by ksco** [reply]
Here is a blog post on how library wrapping works in general: https://box86.org/2021/08/a-deep-dive-into-library-wrapping/.
